lib/zlib/zlib.mk

```diff
@@ -0,0 +1,25 @@
+#
+# Copyright (c) 2018, ARM Limited and Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+ZLIB_PATH	:=	lib/zlib
+
+# Imported from zlib 1.2.11 (do not modify them)
+ZLIB_SOURCES	:=	$(addprefix $(ZLIB_PATH)/,	\
+					adler32.c	\
+					crc32.c		\
+					inffast.c	\
+					inflate.c	\
+					inftrees.c	\
+					zutil.c)
+
+# Implemented for TF
+ZLIB_SOURCES	+=	$(addprefix $(ZLIB_PATH)/,	\
+					gunzip.c)
```

Comment:
The use of gunzip() is not limited to the image decompressor, rather it is a generic helper to decompress arbitrary data.
If we want to prevent platforms from using gunzip() for other purposes, we can guard it with `ifeq($(FIP_IMAGE_COMPRESS),gzip)`.

BTW, `COMPRESS_LOG_algo` does not exit - it only appears in comment lines.  I assume you meant `ifeq($(FIP_IMAGE_COMPRESS),gzip)`.

One more thing, I am following the existing policy.
GENERATE_COT and TRUSTED_BOARD_BOOT are not forced to be pair.
GENERATE_COT only affects host tooling.  If we generate certificates, it is very likely the TF needs the capability to verity them, however TRUSTED_BOARD_BOOT is not forced.
(So, why do we need two options?)

All upstreamed platforms with Bl1/BL2 support use FIP, but theoretically, FIP is just one form of abstraction in the IO layer. (drivers/io/io_fip.c)
Platforms are allowed to handle bare images if they do not want to use FIP.
I am keeping the possibility for platforms to load compressed bare images.

If we want to turn the use of FIP mandatory for BL1 and BL2, (or at least FIP is mandatory for data compression) we can use `FIP_IMAGE_COMPRESS` as the compile switch for TF, but I'd like to know if we really should  do this.

Reply:
I think it's correct this library is a generic helper. I'm not too worried if the platform has not integrated a matching library; the runtime image decompressor should fail gracefully in this case (and the implementation here does). If in future there are several decompression algorithms with separate host and device libraries, we can look to enforce matching at that time.

Regarding GENERATE_COT and TRUSTED_BOARD_BOOT, I agree there isn't a strong reason for these to be separate options. It is probably only that way for legacy reasons.

It is correct to allow the possibility for platforms use to this compression functionality without FIP. Both compression and FIP should be independent platform choices.